File: JavaScript/src/Internals/Functions/Csharp/Function_HASVALUE.js
import { Function_2 } from '../Function_2.js';
import { Operand } from '../../../Operand.js';

/**
 * Function_HASVALUE
 */
export class Function_HASVALUE extends Function_2 {
    /**
     * @param {FunctionBase[]} funcs
     */
    constructor(funcs) {
        super(funcs);
    }
    
    get Name() {
        return "HasValue";
    }
    
    /**
     * @param {AlgorithmEngine} engine
     * @returns {Operand}
     */
    evaluate(engine, tempParameter) {
        let args1 = this.a.evaluate(engine, tempParameter);
        if (args1.IsError) { return args1; }

        let args2 = this.getText_2(engine, tempParameter);
        if (args2.IsError) { return args2; }

        if (args1.IsArrayJson) {
            return Operand.Create(args1.ContainsValue(args2));
        } else if (args1.IsJson) {
            let json = args1.JsonValue;
            if (json.IsArray) {
                for (let i = 0; i < json.Count; i++) {
                    let v = json.inst_array[i];
                    if (v.IsString) {
                        if (v.StringValue === args2.TextValue) {
                            return Operand.True;
                        }
                    } else if (v.IsDouble) {
                        if (v.NumberValue.toString() === args2.TextValue) {
                            return Operand.True;
                        }
                    } else if (v.IsBoolean) {
                        if (v.BooleanValue.toString().toLowerCase() === args2.TextValue.toLowerCase()) {
                            return Operand.True;
                        }
                    }
                }
            } else {
                for (let key in json.inst_object) {
                    let v = json.inst_object[key];
                    if (v.IsString) {
                        if (v.StringValue === args2.TextValue) {
                            return Operand.True;
                        }
                    } else if (v.IsDouble) {
                        if (v.NumberValue.toString() === args2.TextValue) {
                            return Operand.True;
                        }
                    } else if (v.IsBoolean) {
                        if (v.BooleanValue.toString().toLowerCase() === args2.TextValue.toLowerCase()) {
                            return Operand.True;
                        }
                    }
                }
            }
            return Operand.False;
        } else if (args1.IsArray) {
            let ar = args1.ArrayValue;
            for (let item of ar) {
                let t = item.ToText();
                if (t.IsError || t.IsNone) {
                    continue;
                }
                if (t.TextValue === args2.TextValue) {
                    return Operand.True;
                }
            }
            return Operand.False;
        }
        return this.parameterError(1);
    }
    

}

File: java/src/main/java/toolgood/algorithm/internals/functions/csharp/Function_HASVALUE.java
package toolgood.algorithm.internals.functions.csharp;

import java.util.List;
import java.util.function.BiFunction;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.operands.OperandKeyValueList;
import toolgood.algorithm.operands.OperandArray;
import toolgood.algorithm.litJson.JsonData;
import toolgood.algorithm.internals.functions.Function_2;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_HASVALUE extends Function_2 {
    public Function_HASVALUE(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "HasValue";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = func1.Evaluate(engine, tempParameter);
        if (args1.IsErrorOrNone()) { return args1; }

        Operand args2 = GetText_2(engine, tempParameter);
        if (args2.IsErrorOrNone()) { return args2; }

        if (args1.IsArrayJson()) {
            return Operand.Create(((OperandKeyValueList) args1).ContainsValue(args2));
        } else if (args1.IsJson()) {
            JsonData json = args1.JsonValue();
            if (json.IsArray()) {
                for (int i = 0; i < json.Count(); i++) {
                    JsonData v = json.get(i);
                    if (v.IsString()) {
                        if (v.StringValue().equals(args2.TextValue())) { return Operand.True; }
                    } else if (v.IsDouble()) {
                        if (v.NumberValue().toString().equals(args2.TextValue())) { return Operand.True; }
                    } else if (v.IsBoolean()) {
                        if (Boolean.toString(v.BooleanValue()).equalsIgnoreCase(args2.TextValue())) { return Operand.True; }
                    }
                }
            } else {
                for (JsonData v : json.inst_object.values()) {
                    if (v.IsString()) {
                        if (v.StringValue().equals(args2.TextValue())) { return Operand.True; }
                    } else if (v.IsDouble()) {
                        if (v.NumberValue().toString().equals(args2.TextValue())) { return Operand.True; }
                    } else if (v.IsBoolean()) {
                        if (Boolean.toString(v.BooleanValue()).equalsIgnoreCase(args2.TextValue())) { return Operand.True; }
                    }
                }
            }
            return Operand.False;
        } else if (args1.IsArray()) {
            OperandArray ar = (OperandArray) args1;
            for (Operand item : ar.ArrayValue()) {
                Operand t = item.ToText(null);
                if (t.IsErrorOrNone()) { continue; }
                if (t.TextValue().equals(args2.TextValue())) {
                    return Operand.True;
                }
            }
            return Operand.False;
        }
        return ParameterError(1);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.BOOLEAN;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.JSON, null, null);
        func2.GetParameterTypes(noneEngine, result, OperandType.TEXT, null, null);
    }
}
